### ISSUE TITLE
Functionality / doc question: Is there a way in stainless to declare a type as @extern? 

### ISSUE THREAD

**Issue by saraswat** [epfl-lara/stainless#766]
I [see](https://epfl-lara.github.io/stainless/wrap.html) that fields can be marked `@extern` (their type will become opaque, stainless will not descend into the type).

Is there a way to directly declare a _type_ as `@extern` (so that `stainless` is asked not to examine the type)? (In real programs there could be a large number of occurrences of the type, and really you don't want `stainless` to examine any field / variable etc of that type.)

I tried annotating the `import` statement for the type, that did not work.

**Comment by romac** [comment]
Unfortunately there is no way to mark a type as `@extern` at the moment, but that does indeed seem like a valuable idea.

The closest thing to extern types we currently have is the combination of `@extern` field and the auto conversion of unknown types in the signature of `@extern` functions to the aforementioned opaque types. Please find a (somewhat contrived) example below demonstrating both features:

```scala
import stainless.annotation._
import java.security.MessageDigest

object test {

  case class Digest(
    @extern inner: MessageDigest, // <-----
    var reset: Boolean,
  ) {

    @extern
    def update(buffer: java.nio.ByteBuffer): Unit = { // <-----
      inner.update(buffer)
      reset = false
    } ensuring { reset == false }

    @extern
    def update(buffer: Array[Byte]): Unit = {
      inner.update(buffer)
      reset = false
    } ensuring { reset == false }

    @extern
    def digest(): String = {
      val res = inner.digest().toString
      reset = true
      res
    } ensuring { reset == true }
  }

  object Digest {
    @extern @pure
    def sha256: Digest = {
      Digest(MessageDigest.getInstance("SHA-256"), reset = true)
    } ensuring { _.reset == true }
  }

  def test = {
    val d = Digest.sha256
    assert(d.reset)
    d.update(Array[Byte](1,2,3))
    assert(!d.reset)
    d.update(Array[Byte](4,5,6))
    assert(!d.reset)
    val res = d.digest()
    assert(d.reset)
  }
}
```

The only two lines worth looking at are the ones annotated with `<-----`, but I still included the full benchmark as to provide a working example that can be played with.

**Comment by saraswat** [comment]
Also, @romac it seems that this does what I wanted (types being made `@extern`) ....?

Let's say you have an externally defined type `a.b.c.Foo` (defined in a class file in a jar, say). You have lots of code that refers to it, with an `import a.b.c.Foo` before it. Now do this: 

```
// import a.b.c.Foo
@extern type Foo = a.b.c.Foo
```

At least this has worked in a simple example I tried. Should this work, given the design and the implementation of `@extern`...?

**Comment by saraswat** [comment]
This is how your example looks, with this "opaque type declaration" trick, not very different. 

But I have successfully now converted about 200 lines of code from Scala to Stainless using the opaque type declaration trick, with very few lines of change (because the code was just using external types opaquely).

```
import stainless.annotation._
import stainless.lang.{Option, None,Some}
object test {
  @extern type MessageDigest = java.security.MessageDigest
  case class Digest(
    inner: Option[MessageDigest], 
    var reset: Boolean,
  ) {
    @extern
    def update(buffer: java.nio.ByteBuffer): Unit = { 
      inner.get.update(buffer)
      reset = false
    } ensuring { reset == false }

    @extern
    def update(buffer: Array[Byte]): Unit = {
      inner.get.update(buffer)
      reset = false
    } ensuring { reset == false }

    @extern
    def digest(): String = {
      val res = inner.get.digest().toString
      reset = true
      res
    } ensuring { reset == true }
  }

  object Digest {
    @extern
    def sha256: Digest = {
      Digest(Some(java.security.MessageDigest.getInstance("SHA-256")), reset = true)
    } ensuring { _.reset == true }
  }

  def test = {
    val d = Digest.sha256
    assert(d.reset)
    d.update(Array[Byte](1,2,3))
    assert(!d.reset)
    d.update(Array[Byte](4,5,6))
    assert(!d.reset)
    val res = d.digest()
    assert(d.reset)
  }
}
```

**Comment by romac** [comment]
Right! That's indeed by design, I had just somehow forgotten about `@extern` type aliases, sorry about that. Glad to see you've made good use of that feature, and thanks for the updated example :)

**Comment by romac** [comment]
> This brings up the question -- is it possible to decorate a statement with `@extern` (or, perhaps, more appropriately, `@opaque`)?

Not to my knowledge, but that could be a nice feature to have!
Should we have it, I would use `@extern` instead of `@opaque` since the latter already has a specific meaning: it hides the body of the annotated function to the solver when solving VCs other than its own.

**Comment by romac** [comment]
> Also I notice that I can mark classes / objects as @extern. But Stainless does seem to descend into them. So what exactly does @extern do in this case? (My theory is that every method in its body is marked @extern.)

Yes that's indeed what's happening!
